### DISCUSSION TITLE
Helping newcomers with ISO verification

### DISCUSSION THREAD

**Discussion by spamegg1** [linuxmint/discussions#674] [Ideas]
There are many newcomers from Windows who have no idea how to use the Terminal, or what the commands mean.

Even with a nicely written documentation like [Mint install docs](https://forums.linuxmint.com/viewtopic.php?t=434001) some newcomers struggle, for [example](https://forums.linuxmint.com/viewtopic.php?t=434001). There are many more examples.

Of course we can say that "well they have to learn at some point and get used to it", but some newcomers simply leave, or get wrong results because they are in the wrong directory / some file is missing, or just try installing unverified ISOs and run into problems during installation (because the download did not complete fully / correctly, etc.)

Currently the nice GUI verification available on Mint uses `mint-iso-verify`, which uses the `verify.py` file in `/usr/bin/mintstick`:

```bash
❯ bat /usr/bin/mint-iso-verify
───────┬───────────────────────────────────────────────────────────────────────────────
       │ File: /usr/bin/mint-iso-verify
───────┼───────────────────────────────────────────────────────────────────────────────
   1   │ #!/bin/sh
   2   │ #
   3   │ set -e
   4   │ 
   5   │ TERM=xterm
   6   │ 
   7   │ DIR=/usr/lib/mintstick
   8   │ 
   9   │ $DIR/verify.py "$@"
───────┴─────────────────────
```

How easy / difficult would it be to make `mintstick` + `verify.py` cross-platform so that it could be used on Windows and Mac directly, or in the case of the above example, easily installable / usable on other Linux distros? So these users can verify graphically without the Terminal. I suppose it also depends on `gpg` and `sha256sum`. Maybe it's more difficult than I think?

`mintstick` can be installed on Ubuntu by downloading the `.deb` package from [Mint repo](http://packages.linuxmint.com/) but does it come with `verify.py` and the `/usr/bin/mint-iso-verify` script after installation?

And I'm aware that even if this is possible, `mintstick` itself would have to be trusted first... :confused:

**Comment by 100savage** [comment]
Look in the System Settings -> Actions -> Download. You will see several ways to verify an ISO

**Reply by spamegg1** [reply]
I'm talking about making it work on Windows / Mac / other Linux distros.

**Reply by Jopp-gh** [reply]
On Unix systems, like MacOs and Linux, you can use the same bash commands to verify hash checksums. 

So yes, it's possible to create a bash script, put into a wrapper and check downloaded ISO files on both Mac and Linux. 

Win instead, uses power shell which is isn't portable and only a win thing. 
Maybe a python script could help? I don't use python so idk.

**Reply by spamegg1** [reply]
Linux + Mac come with some version of Python installed but Windows does not. So the script would have to be ported to Powershell I guess...

**Reply by Jopp-gh** [reply]
You don't need python for checksums, which is the main part of the script. Also, you can convert windows paths to POSIX compliant paths and vice versa. 

Btw, Powershell includes checksum options, how a quick look at stackoverflow shows:  https://stackoverflow.com/questions/63396620/how-can-i-compare-a-files-sha256-hash-in-powershell-to-a-known-value#63396621 

It would be pretty straightforward to create a small, easy to use drag n drop batch script and submit it to the devs of Mint for approval. I don't have Windows so I can't do it.

**Reply by 100savage** [reply]
Here is a powershell script you can use
```# Path to the ISO file
$filePath = "C:\Path\To\Your\File.iso"

# Calculate the hash using SHA256 algorithm
$fileHash = Get-FileHash -Path $filePath -Algorithm SHA256

# Display the file hash
Write-Host "File Path: $($fileHash.Path)"
Write-Host "Hash: $($fileHash.Hash)"
Write-Host "Algorithm: $($fileHash.Algorithm)"
```
Replace "C:\Path\To\Your\File.iso" with the actual path to your ISO file. This script will compute and display the hash of the specified file using the SHA256 algorithm. You can change the algorithm to MD5 or SHA1 if needed by modifying the -Algorithm parameter.

Does this help?

**Reply by Jopp-gh** [reply]
If you save this code as a batch file, users don't have to add file paths and can drag n drop - verify ISO files in place. 
If drag n drop isn't possible, you should be able to add generic paths like home or download. 

> [!NOTE] 
> However, there's a fundamental question:
How many users know and follow good practices like to verify ISO  files ?

I mean, ofc you better check, but probably the average user will download Mint from [the main page of Mint](https://www.linuxmint.com/download.php), reboot from USB stick, have a first look at Mint and eventually, install Mint.

**Comment by Qronikarz** [comment]
Tails documentation has a great article about .iso verification. They also use JavaScript to allow the user to verify the .iso
https://tails.net/contribute/design/download_verification/

No windows user will run a python, shell or anything like that to verify Mint .iso. The moment you need your user to touch the terminal/console to do anything you lose most of them.

**Comment by spamegg1** [comment]
After thinking about this more, I realize how challenging and complicated it is. I'll leave it for now.
